modulus/models/gnn_layers/bsms.py

```diff
@@ -0,0 +1,582 @@
+            pos = self.edge_conv(pos, m_gs[i], ew)
+            cts.append(ew)
+
+            # Pooling
+            if len(h.shape) == 3:
+                h = h[:, m_ids[i]]
+            elif len(h.shape) == 2:
+                h = h[m_ids[i]]
+
+            if len(pos.shape) == 3:
+                pos = pos[:, m_ids[i]]
+            elif len(pos.shape) == 2:
+                pos = pos[m_ids[i]]
+
+            w = w[m_ids[i]]
+
+        # Bottom pass
+        h = self.bottom_gmp(h, m_gs[self.unet_depth], pos)
+
+        # Up pass
+        for i in range(self.unet_depth):
+            depth_idx = self.unet_depth - i - 1
+            g, idx = m_gs[depth_idx], m_ids[depth_idx]
+            h = self.unpools[i](h, down_outs[depth_idx].shape[-2], idx)
+            # aggregate is False as we are returning the information to previous out degrees.
+            h = self.edge_conv(h, g, cts[depth_idx], aggragating=False)
+            h = self.up_gmps[i](h, g, down_ps[depth_idx])
+            h = h.add(down_outs[depth_idx])
+
+        return h
+
+
+class GraphMessagePassing(nn.Module):
+    """Graph Message Passing (GMP) block."""
+
+    def __init__(self, latent_dim, hidden_layer, pos_dim):
+        """
+        Initialize the GMP block.
+
+        Parameters
+        ----------
+        latent_dim : int
+            Dimension of the latent space.
+        hidden_layer : int
+            Number of hidden layers.
+        pos_dim : int
+            Dimension of the positional encoding.
+        """
+        super().__init__()
+        self.mlp_node = MeshGraphMLP(
+            2 * latent_dim, latent_dim, latent_dim, hidden_layer
+        )
+        edge_info_in_len = 2 * latent_dim + pos_dim + 1
+        self.mlp_edge = MeshGraphMLP(
+            edge_info_in_len, latent_dim, latent_dim, hidden_layer
+        )
+        self.pos_dim = pos_dim
+
+    def forward(self, x, g, pos):
+        """
+        Forward pass for GMP block.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input node features of shape [B, N, C] or [N, C].
+        g : torch.Tensor
+            Graph connectivity (edges) of shape [2, E].
+        pos : torch.Tensor
+            Node positional information of shape [B, N, pos_dim] or [N, pos_dim].
+
+        Returns
+        -------
+        torch.Tensor
+            Updated node features.
+        """
+        i, j = g[0], g[1]
+
+        if len(x.shape) == 3:
+            B, _, _ = x.shape
+            x_i, x_j = x[:, i], x[:, j]
+        elif len(x.shape) == 2:
+            x_i, x_j = x[i], x[j]
+        else:
+            raise ValueError(f"Only implemented for dim 2 and 3, got {x.shape}")
+
+        if len(pos.shape) == 3:
+            pi, pj = pos[:, i], pos[:, j]
+        elif len(pos.shape) == 2:
+            pi, pj = pos[i], pos[j]
+        else:
+            raise ValueError(f"Only implemented for dim 2 and 3, got {x.shape}")
+
+        # NOTE for zijie
```

Comment:
Was this left intentionally?

Reply:
Oops, yes, thank you - I've removed that note comment and cleaned up the comment itself.